content/6_Advanced/DP_SOS.mdx

```diff
@@ -26,6 +26,85 @@ frequency: 0
 	</Resource>
 </Resources>
 
+
+Sum over subsets (SOS) dp is a trick that allows you to efficiently compute the sum of all the subsets of an array.
+
+The naive solution would be to iterate through every other mask and if the other mask is a subset of the current mask we add it to the current answer.
+
+<LanguageSection>
+<CPPSection>
+```cpp
+int n = 20;
+vector<int> a(1 << n);
+vector<int> sos(1 << n);
+for (int i = 0; i < (1 << n); i++) {
+	for (int j = 0; j < (1 << n); j++) {
+		// j is a subset of i
+		if ((i & j) == j) { sos[i] += a[j]; }
+	}
+}
+```
+</CPPSection>
+</LanguageSection>
+
+
+We can speed this up if we iterate over only subsets of the current mask and add up all of the those values to get the sum over subsets for a particular mask.
+The time complexity here is O($3^n$)
+
+<LanguageSection>
+<CPPSection>
+```cpp
+int n = 20;
+vector<int> a(1 << n);
+vector<int> sos(1 << n);
+for (int i = 0; i < (1 << n); i++) {
+	// Iterates over all subsets of i
```

Comment:
```suggestion
	// iterate over all subsets of i
```
why not use the same technique you did in the above code snippet

Reply:
It's faster, O(3^n) as opposed to O(4^n)

Reply:
you're still iterating over all subsets
how is one method faster than the other
they go over the same subsets(?)

Reply:
One of them directly iterates over all subsets of the mask while the other iterates over every other mask and checks whether it is a submask the original mask. I will clarify that ig.